### PR TITLE
Issue 63: Improve pravega-provisioner tool to use it from other scripts

### LINKED ISSUE #63: improve pravega-provisioner tool to use it from other scripts
pravega-provisioner could be used other scripts to get pravega cluster sizing recommendations.
When pravaga provisioner is launched from helper scripts to get recommended configuration
it prompts for user input. If we want to controller the input needed from end user, it would be
better to parameterize methods for a given provisioning model.
Similarly, if output is returned as python object (a dictionary) with well known keys it would be easy to consume instead of parsing the text output printed by provisioner.
Further we may want to control the messages emitted to screen, as unwanted messages will clutter the screen and confuse the end user.
Input errors are not handled, instead of baling out it should prompt for input again.

### PR DESCRIPTION
**Pull request by gaddamas**
### Purpose of the change
Fixes #63

### What the code does
1. This PR aims to parameterize input to resource_based_provisioning, so that calling script can gather required input from end user in customized way and pass it to the model.
2. Returns output as python object (a dictionary) with well known keys.
3. Prints messages to screen based on verbose setting.
4. Fixes input methods to retry instead of bailing out.
5. Fixes python import errors.

### How it is verified

Ran the script directly and also called from another python program.

1) Direct invocation
```
python3 ./pravega-provisioner/cluster_provisioner.py
### Provisioning model for Pravega clusters ###
Please, introduce the resource information about the VMs used int cluster:
How many CPU cores has each VM/node?72
How much memory in GB has each VM/node?384
Is your cluster using local drives? (yes, no)6
Is your cluster using local drives? (yes, no)yes
How many local drives has each VM/node?6
Do you want to provision a Pravega cluster based on 0) the resources available, 1) the input workload, 2) none?(valid values: [0, 1, 2])0
How many VMs/nodes do you want to devote for a Pravega cluster?4
How many node failures you want to tolerate?(valid values: range(0, 100))0
Is the workload metadata-heavy (i.e., many clients, small transactions)? (yes, no)no
Allocation of pods on nodes:  [(22, 284, 0, ['bk', 'bk', 'bk', 'ss', 'ss', 'ss', 'c']), (21, 282, 0, ['bk', 'bk', 'bk', 'ss', 'ss', 'ss', 'c', 'zk']), (21, 282, 0, ['bk', 'bk', 'bk', 'ss', 'ss', 'ss', 'c', 'zk']), (21, 282, 0, ['bk', 'bk', 'bk', 'ss', 'ss', 'ss', 'c', 'zk'])]
--------- Segment Store In-Memory Cache Size (Pravega +0.7) ---------
Segment Store pod memory limit:  110 GB
Segment Store JVM Size (-Xmx JVM Option) :  4 GB
Segment Store direct memory (-DirectMemory JVM Option):  106 GB
Segment Store cache size (pravegaservice.cacheMaxSize):  112742891520  ( 105 GB)
Buffering time that Pravega tolerates with Tier 2 unavailable given a (well distributed) write workload:
- Write throughput:  100 (MBps) -> buffering time:  12902.4  seconds
- Write throughput:  300 (MBps) -> buffering time:  4300.8  seconds
- Write throughput:  500 (MBps) -> buffering time:  2580.48  seconds
- Write throughput:  700 (MBps) -> buffering time:  1843.2  seconds
- Write throughput:  900 (MBps) -> buffering time:  1433.6  seconds
WARNING: To guarantee data availability and durability, consider enabling rack-aware placement in Pravega to write to Bookkeeper.
--------- Cluster Provisioning ---------
Number of VMs required:  4
Number of Zookeeper servers required:  3
Number of Bookkeeper servers required:  12
Number of Segment Stores servers required:  12
Number of Controller servers required:  4
--------- Cluster Config Params ---------
Number of Segment Containers in config:  96
Number of Stream Buckets in config:  16
```

2)  calling it from another python program (nautilus-dist/scripts/provisioner.py)
https://eos2git.cec.lab.emc.com/NAUT/nautilus-dist/pull/294

NOTE: --num-cpu entered below is physical cores, inside provisioner.py it is multiplied by 2 (vCPU:pCPU ration assumed to be 2:1)
```
 python3 ./provisioner.py --outfile provisioner_output.txt \
     --num-hosts-present 4 \
     --num-hosts-added 2 \
     --num-host-failures 0 \
     --num-cpu 36 \
     --mem 384 \
     --local-disk-count 6 \
     --is-bk-on-bosh True \ 
     --percent-analytics 50 

Gathering input required to scaling the cluster:

Percent of resources for Analytics 50
Percent of resources for Pravega 40
Resources for pravega per each new ESXi host: vCPU: 24, memory: 160 GB
--------- Segment Store In-Memory Cache Size (Pravega +0.7) ---------
Segment Store pod memory limit:  128 GB
Segment Store JVM Size (-Xmx JVM Option) :  4 GB
Segment Store direct memory (-DirectMemory JVM Option):  124 GB
Segment Store cache size (pravegaservice.cacheMaxSize):  132070244352  ( 123 GB)
Buffering time that Pravega tolerates with Tier 2 unavailable given a (well distributed) write workload:
- Write throughput:  100 (MBps) -> buffering time:  2519.04  seconds
- Write throughput:  300 (MBps) -> buffering time:  839.68  seconds
- Write throughput:  500 (MBps) -> buffering time:  503.81  seconds
- Write throughput:  700 (MBps) -> buffering time:  359.86  seconds
- Write throughput:  900 (MBps) -> buffering time:  279.89  seconds
WARNING: To guarantee data availability and durability, consider enabling rack-aware placement in Pravega to write to Bookkeeper.
--------- Segment Store In-Memory Cache Size (Pravega +0.7) ---------
Segment Store pod memory limit:  32 GB
Segment Store JVM Size (-Xmx JVM Option) :  4 GB
Segment Store direct memory (-DirectMemory JVM Option):  28 GB
Segment Store cache size (pravegaservice.cacheMaxSize):  28991029248  ( 27 GB)
Buffering time that Pravega tolerates with Tier 2 unavailable given a (well distributed) write workload:
- Write throughput:  100 (MBps) -> buffering time:  1935.36  seconds
- Write throughput:  300 (MBps) -> buffering time:  645.12  seconds
- Write throughput:  500 (MBps) -> buffering time:  387.07  seconds
- Write throughput:  700 (MBps) -> buffering time:  276.48  seconds
- Write throughput:  900 (MBps) -> buffering time:  215.04  seconds

Suggested additional worker VM count for pravega: 10
Suggested additional pravega replicas:
  zk instaces: 3
  bk instances on Bosh: 3
  segment stores: 7
  controllers: 3
```

### CODE REVIEW

File: pravega_provisioner/cluster_provisioner.py [hunk 1]
```diff
@@ -19,11 +20,16 @@ def get_bool_input(msg):
     :param msg: Message to be prompted to the user.
     :return: Valid user input.
     """
-    user_input = input(msg + " (yes, no)")
-    while user_input != "yes" and user_input != "no":
-        user_input = input(msg + " (yes, no)")
-    return user_input == "yes"
-
+    return_value = False
```
Comment (RaulGracia) [inline pravega_provisioner/cluster_provisioner.py]: what is the point of this change? If this is a problem related to using `raw_input` instead of `input`? This works for me using `Python 3.7.2`, so if it is not a necessary change, please revert it.

File: pravega_provisioner/cluster_provisioner.py [hunk 2]
```diff
@@ -257,8 +302,19 @@ def main():
 
     if provisioning_model == 0:
         vms = int(input("How many VMs/nodes do you want to devote for a Pravega cluster?"))
-        zookeeper_servers, bookkeeper_servers, segment_stores, controllers = resource_based_provisioning(vms, vm_cpus,
-                        vm_ram_gb, vm_local_drives, zookeeper_servers, bookkeeper_servers, segment_stores, controllers)
+        # Ask to the user the number of node failures he/she wants to tolerate.
+        failures_to_tolerate = get_int_input("How many node failures you want to tolerate?", range(0, 2))
```
Comment (RaulGracia) [inline pravega_provisioner/cluster_provisioner.py]: Please, leave the range of failures to what it was, as people may want to deploy Pravega on huge environments in which experiencing more than 2 failures is a common scenario.

File: pravega_provisioner/cluster_provisioner.py [hunk 3]
```diff
@@ -154,32 +181,50 @@ def resource_based_provisioning(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookee
     # Get the resources used for the last possible allocation
     the_cluster = can_allocate_services_on_nodes(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookeeper_servers,
                                                  bookkeeper_servers, segment_stores, controllers)[1]
-    print("Allocation of pods on nodes: ", the_cluster)
+    if verbose:
```
Comment (RaulGracia) [inline pravega_provisioner/cluster_provisioner.py]: Please, remove all the `if verbose`, this makes the code more complex without need. The script may print things anyway, I don't see why it should not (it can be used anyway from another program).

File: pravega_provisioner/model/provisioning_logic.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 #
 #     http://www.apache.org/licenses/LICENSE-2.0
 #
-from model.constants import Constants
+from constants import Constants
```
Comment (RaulGracia) [inline pravega_provisioner/model/provisioning_logic.py]: This does not work, please revert this change:
```
λ python cluster_provisioner.py
Traceback (most recent call last):
  File "cluster_provisioner.py", line 13, in <module>
    from model.provisioning_logic import *
  File "xxx\pravega-tools\pravega_provisioner\model\provisioning_logic.py", line 9, in <module>
    from constants import Constants
ModuleNotFoundError: No module named 'constants'
```

**Review by RaulGracia — CHANGES_REQUESTED** [review]
I have several comments at first glance. Once you address them, I will give more. Also, apart from these comments, rever the naming change from `pravega-provisioner` to `pravega_provisioner`, as it is unnecessary.

File: pravega_provisioner/cluster_provisioner.py (hunk 2 above)
Comment (gaddamas) [reply pravega_provisioner/cluster_provisioner.py]: Can we change to 10 instead of original value 100. I think Ashish also asked to reduce this.
please see comment 1 in https://asdwiki.isus.emc.com:8443/display/NAUT/Pravega+Cluster+Provisioning+Model

Comment (RaulGracia) [reply pravega_provisioner/cluster_provisioner.py]: I saw the comment, and I don't see the point of limiting the script to 2 or 10 failures. Is just a modeling script, no something related to any infrastructure.

File: pravega_provisioner/cluster_provisioner.py (hunk 1 above)
Comment (gaddamas) [reply pravega_provisioner/cluster_provisioner.py]: done. yes it worked with python3.

File: pravega_provisioner/cluster_provisioner.py (hunk 3 above)
Comment (gaddamas) [reply pravega_provisioner/cluster_provisioner.py]: done. Also I rounded the decimals to 2.
```
- Write throughput:  100 (MBps) -> buffering time:  6615.04  seconds
- Write throughput:  300 (MBps) -> buffering time:  2205.01  seconds
- Write throughput:  500 (MBps) -> buffering time:  1323.01  seconds
- Write throughput:  700 (MBps) -> buffering time:  945.01  seconds
- Write throughput:  900 (MBps) -> buffering time:  735.0  seconds
```
old output:
```
- Write throughput:  100 (MBps) -> buffering time:  6615.04  seconds
- Write throughput:  300 (MBps) -> buffering time:  2205.0133333333333  seconds
- Write throughput:  500 (MBps) -> buffering time:  1323.008  seconds
- Write throughput:  700 (MBps) -> buffering time:  945.0057142857142  seconds
- Write throughput:  900 (MBps) -> buffering time:  735.0044444444444  seconds
```

File: pravega_provisioner/model/provisioning_logic.py (hunk 4 above)
Comment (gaddamas) [reply pravega_provisioner/model/provisioning_logic.py]: I need to figure out what to do, when using pravega-provisioner as module from scaling script it is not able to find constants.

File: pravega_provisioner/cluster_provisioner.py (hunk 1 above)
Comment (gaddamas) [reply pravega_provisioner/cluster_provisioner.py]: Reverted dir name change from pravega-provisioner to pravega_provisioner.
I figured our a way to pravega-provisioner as module,  with dash in name python is complaining, while downloading sources to scripts distribution I am copying to directory pravega_provisioner

File: pravega_provisioner/model/provisioning_logic.py (hunk 4 above)
Comment (RaulGracia) [reply pravega_provisioner/model/provisioning_logic.py]: @gaddamas please let me know when you fix this issue to try this PR again.

Comment (gaddamas) [reply pravega_provisioner/model/provisioning_logic.py]: Yes, I fixed it. reverted change to import statement.
Please see files changed and see the diff.

File: pravega-provisioner/cluster_provisioner.py [hunk 5]
```diff
@@ -161,25 +183,40 @@ def resource_based_provisioning(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookee
...
-    print("Segment Store JVM Size (-Xmx JVM Option) : ", Constants.segment_store_jvm_size_in_gb, "GB")
-    print("Segment Store direct memory (-DirectMemory JVM Option): ", new_direct_memory, "GB")
-    # We leave 1GB extra of direct memory for non-caching purposes
-    print("Segment Store cache size (pravegaservice.cacheMaxSize): ", (new_direct_memory - 1) * 1024 * 1024 * 1024,
-          " (", new_direct_memory - 1, "GB)")
-    print("Buffering time that Pravega tolerates with Tier 2 unavailable given a (well distributed) write workload:")
-    for w in range(100, 1000, 200):
-        print("- Write throughput: ", w, "(MBps) -> buffering time: ", ((segment_stores * (new_direct_memory - 1) * 1024) / w), " seconds")
+    new_direct_memory = 0 
+    if max_segment_stores_per_vm > 0:
```
Comment (RaulGracia) [inline pravega-provisioner/cluster_provisioner.py]: `max_segment_stores_per_vm` should be always > 0, given that if not the program will exit in the previous assertion that checks for the allocation of a "minimal cluster" (that requires at least 1 segment store). So you can remove this if block.

File: pravega-provisioner/cluster_provisioner.py [hunk 6]
```diff
@@ -38,6 +39,30 @@ def get_int_input(msg, valid_values):
     return user_input
 
 
+def get_int_input_with_default(msg, valid_values, default_value):
```
Comment (RaulGracia) [inline pravega-provisioner/cluster_provisioner.py]: I don't see this method being used. If so, please remove it.

File: pravega-provisioner/cluster_provisioner.py [hunk 7]
```diff
@@ -161,25 +183,40 @@ def resource_based_provisioning(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookee
...
+            print("- Write throughput: ", w, "(MBps) -> buffering time: ", str(round((segment_stores * (new_direct_memory - 1) * 1024) / w, 2)), " seconds")
 
     # Add a warning if the number of Bookies is higher than the number of nodes, as Pravega may need to enable rack-
     # aware placement in Bookkeeper so it tries to write to Bookies in different nodes (data availability).
     if bookkeeper_servers > vms:
         print("WARNING: To guarantee data availability and durability, consider enabling rack-aware placement in "
               "Pravega to write to Bookkeeper.")
 
-    return zookeeper_servers, bookkeeper_servers, segment_stores, controllers
+    # zookeeper_servers, bookkeeper_servers, segment_stores, controllers
```
Comment (RaulGracia) [inline pravega-provisioner/cluster_provisioner.py]: It is fine to use a dict, but you can remove the comment justifying that decision; just change it.

File: pravega-provisioner/cluster_provisioner.py [hunk 8]
```diff
@@ -161,25 +183,40 @@ def resource_based_provisioning(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookee
...
+    # return values in a dictionary instead of relying on long list of positional output values
+    # this also allows in adding any new values if required
+    output = dict()
+    output['controller'] = int(controllers)
```
Comment (RaulGracia) [inline pravega-provisioner/cluster_provisioner.py:187]: Please, if you are adding the output, please add all the fields (not only the ones that are good for the use that you want to do of this script). Please, this is an open source project, so let's think on the completeness of the code and genera use. For instance, why you are not adding the number of nodes/vms here?

File: pravega-provisioner/model/provisioning_logic.py [hunk 9]
```diff
@@ -183,7 +183,7 @@ def _subtract_resources_from_cluster(the_cluster, num_instances, cpu_per_instanc
     finish_allocation = False
     completed_allocations = retries = 0
     while not finish_allocation:
-        for x in range(num_instances - completed_allocations):
+        for x in range(int(num_instances) - completed_allocations):
```
Comment (RaulGracia) [inline pravega-provisioner/model/provisioning_logic.py:186]: Is this casting required?

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Some comments.

File: pravega-provisioner/cluster_provisioner.py (hunk 6 above)
Comment (gaddamas) [reply pravega-provisioner/cluster_provisioner.py]: done

File: pravega-provisioner/model/provisioning_logic.py (hunk 9 above)
Comment (gaddamas) [reply pravega-provisioner/model/provisioning_logic.py:186]: yes, somewhere, it is set to floating point.

File: pravega-provisioner/cluster_provisioner.py (hunk 8 above)
Comment (gaddamas) [reply pravega-provisioner/cluster_provisioner.py:187]: number of nodes/vms is input to this method. Previously it was prompted in the method, now it is prompted before calling this method and passed as parameter.

File: pravega-provisioner/cluster_provisioner.py [hunk 10]
```diff
@@ -154,7 +151,8 @@ def resource_based_provisioning(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookee
     # Get the resources used for the last possible allocation
     the_cluster = can_allocate_services_on_nodes(vms, vm_cpus, vm_ram_gb, vm_local_drives, zookeeper_servers,
                                                  bookkeeper_servers, segment_stores, controllers)[1]
-    print("Allocation of pods on nodes: ", the_cluster)
+    if Constants.VERBOSE_OUTPUT:
```
Comment (RaulGracia) [inline pravega-provisioner/cluster_provisioner.py:154]: Please, let's keep this showing the allocation always if it is not a problem.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
One minor comment and one question: why we do not put all the parameters that are shown in the screen in the output dictionary?

**Review by gaddamas — COMMENTED** [review]
added VMS to result dictionary.

File: pravega-provisioner/cluster_provisioner.py (hunk 8 above)
Comment (gaddamas) [reply pravega-provisioner/cluster_provisioner.py:187]: added other variables to output, please see above.

**Review by RaulGracia — APPROVED** [review]
